### DISCUSSION TITLE
Hardware periodically going to sleep?

### DISCUSSION THREAD

**Discussion by rustyhealey** [Dakkaron/Fairberry#27] [General]
I have two questions which fall on the software / configuration side of things:

1. Periodically (sometimes hours, sometimes minutes) the Fairberry hardware will switch modes. The physical keyboard will stop responding and the OSC will kick in. The phone will show a USB device connected and charging, which is normal. If I enter the settings for the USB device and switch "USB controlled by" from "Connected device" to "This device" then it'll all start working again. Normally it's set to "Connected device" and when I attempt to switch the setting, it doesn't actually switch - but the act of trying reactivates the hardware. I'm assuming this is some kind of power save / driver issue. Any idea how to fix this?

2. Periodically I will get a prompt from Android asking which application I want to handle my new USB device. This happens when I first connect the Fairberry, and then some of the times when the first issue happens. I suspect this issue will go away if the first issue is addressed. Any idea what this is? 

Some other observations after having run my S23+ with the Fairberry connected for a half-day:
- Battery life is impacted, but not in a major way. I will keep an eye on this today as I run connected for a full day.
- I had to wirelessly charge my phone upside-down on my Google stand in order to line up contacts. A horizontal pad would fix this.
- Having a physical keyboard is GREAT and Android is perfectly happy with it.
- No major glitches or impediments seen yet to daily driving a phone with Fairberry as primary keyboard.

![IMG_20241217_1508143](https://github.com/user-attachments/assets/d0d8bb24-bfc5-4260-a96a-9a9e8239a437)

https://github.com/user-attachments/assets/474be60e-9a0d-4291-9443-c7625d5dc88f

**Comment by Dakkaron** [comment]
Nice build! Cool that it works well!

Regarding 1: I haven't had that one except when using POWERSAVE_ARDUINO_POWERDOWN in configuration.h. What power saving options do you have enabled in this file? Also, I've seen something similar when the USB connector isn't placed correctly and makes only intermittant connections that beak when the keyboard is moved slightly.

Regarding 2: This happens if you connect the keyboard while having an app installed that can handle USB Serial connections. Yes, it's connected to 1, because 1 acts like reconnecting the keyboard. The other way to go here is to uninstall all apps that can handle USB serial.

Btw, have you checked out this page? https://github.com/Dakkaron/Fairberry/blob/main/Documentation/UX_Shortcuts_and_Apps.md

Regarding the battery consumption: The presence of the Arduino USB serial port causes the USB driver on some phones to stay active continuously, effectively preventing the phone from going into deep sleep. That's by far the biggest impact on battery life. I don't have a solution yet, since Arduino doesn't have a mechanism to disable that USB serial port.

The only workaround I have so far is POWERSAVE_ARDUINO_POWERDOWN which powers down the Atmega and thus disables the USB port, allowing the USB driver to sleep. That then does the same as disconnecting the keyboard, so it needs to reconnect when turning it back on, which causes a delay and will also show that popup you are seeing.

A correct solution would be some way to remove the USB serial port without impacting the USB HID endpoint used by the keyboard.
